### PR TITLE
93 splitting data into sets

### LINKED ISSUE #93: Developing a training pipeline to handle data training, testing, and evaluation and assessment of model performance with K-fold cross-validation
### Motivation

* [x] Data samples are 20 images and labels from [mobius dataset within the github repo](https://github.com/oocular/ready/tree/main/data/mobious/sample-frames/test3000x1700) to accelerate model development. DONE for @mxochicale by creating branch 93-adding-extra-samples
* [ ] Splitting data into train, test and evaluation to develop an appropriate training pipeline. IN PROGRESS for @stoppedtime24 
  * [x] Step 1. Adding Training and validation into the main loop as in this blog https://stanford.edu/~shervine/blog/pytorch-how-to-generate-data-parallel. 
  * [x] Step 2. Adding evaluation with testset as in https://machinelearningmastery.com/training-a-pytorch-model-with-dataloader-and-dataset/
  * [x] Step 3. Loading trained model and evaluate model as shown here https://docs.pytorch.org/tutorials/beginner/saving_loading_models.html?highlight=eval
  * [ ] Review this reference https://debuggercafe.com/saving-and-loading-the-best-model-in-pytorch/
  * Other references: 
    * See how these functions def make_train_dataset, def make_val_dataset, def make_test_dataset are defined here:   https://www.kaggle.com/code/suchith0312/pytorch-dataloader-testing
    * [`train_dataset, test_dataset = torch.utils.data.random_split(full_dataset, [0.8, 0.2])`](https://stackoverflow.com/questions/50544730/how-do-i-split-a-custom-dataset-into-training-and-test-datasets)
  * This might be a clear example 
* [ ] Using K-fold cross validation to asses model performance. See this [reference](https://discuss.pytorch.org/t/using-k-fold-cross-validation-to-train-my-model/196288/2) for `ConcatDataset` and `sklearn.model_selection import KFold`
* [ ] Train model with A100 Nvidia GPU (80GB)
* [ ] Run inference in a local GPU-model (maybe not)?  

### Additional context

* Current [train_unet_with_mobious](https://github.com/oocular/ready/blob/main/src/ready/apis/train_mobious.py) use the whole dataset of 1143 samples of the dataset using [MobiousDataset](https://github.com/oocular/ready/blob/main/src/ready/utils/datasets.py)
* Using [cross validation](https://stackoverflow.com/questions/62954433/how-to-predict-with-the-test-dataset-while-using-cross-validation) for model performance assessment is one good initial step but also you can use it for model tuning (i.e. hyperparameter search).
* Reference describing k-fold-cross-validation:
  * https://www.aionlinecourse.com/tutorial/machine-learning/k-fold-cross-validation
  * https://knowledge.dataiku.com/latest/ml-analytics/ml-concepts/concept-model-validation.html
  * https://stackoverflow.com/questions/58391295/how-is-cross-validation-performed-and-how-gridsearchcv-specifically
  * https://www.mdpi.com/2072-4292/17/7/1256
  * https://www.frontiersin.org/journals/bioengineering-and-biotechnology/articles/10.3389/fbioe.2025.1580502/full
  * Ensembling, Blending & Stacking: https://www.youtube.com/watch?v=TuIgtitqJho

## Notes.
* ...

### PR DESCRIPTION
**Pull request by stoppedtime24**
## Description
Fixes #93 : Split dataset into separate training, validation, and test sets. Converted high-res pics to a lower quality to test pipeline. Added validation into training loop.

## Type of change
Please delete options accordingly to the description.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Suggested Checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation 
- [x] My changes generate no new warnings 
- [x] I have commented my code, particularly in hard-to-understand areas 
- [ ] I have read the [`CONTRIBUTING`](https://github.com/UCL/ready/blob/main/CONTRIBUTING.md) docs
- [x] My code passes, following the style guidelines `pre-commit run -a` 
- [ ] My code is properly tested with `pytest -sv tests`
- [x] I have requested review to this PR.
- [ ] I have addressed and marked as resolved all the review comments in my PR.
- [ ] Make sure your branch is up-to-date with main branch. See step 4 in [`CONTRIBUTING`](https://github.com/UCL/ready/blob/main/CONTRIBUTING.md#pull-request-pr-and-merge-to-main-branch).
- [ ] Finally, I have selected `squash and merge`

### CODE REVIEW

File: data/bash-scripts/resizing_images_in_folder.bash [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+# Create output directory for resized images
+OUTPUT_DIR="$DIR/resized"
+mkdir -p "$OUTPUT_DIR"
+
+# Supported image extensions
+extensions=("jpg" "jpeg" "png")
+
+echo "Resizing images in directory: $DIR"
+echo "Output directory: $OUTPUT_DIR"
+echo "Target size: 576x320"
```
Comment (mxochicale) [inline data/bash-scripts/resizing_images_in_folder.bash:24]: Target size is good, maybe instead of specific target 576x320, can we be more generic using height and width?

File: data/bash-scripts/resizing_images_in_folder.bash [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+# Process each image type
+for ext in "${extensions[@]}"; do
+    # Process both lowercase and uppercase extensions
+    for file in "$DIR"/*."$ext" "$DIR"/*."${ext^^}"; do
+        if [ -f "$file" ]; then
+            filename=$(basename "$file")
+            echo "Processing: $filename"
+            
+            # Resize image maintaining aspect ratio, then crop/pad to exact dimensions
+            convert "$file" -resize 576x320^ -gravity center -extent 576x320 "$OUTPUT_DIR/$filename"
```
Comment (mxochicale) [inline data/bash-scripts/resizing_images_in_folder.bash:36]: Maybe you can add variables for height and width but leave to you to add them or not?

File: data/bash-scripts/resizing_images_in_folder.bash [hunk 3]
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
+
+# Script to resize all images in a directory to 640x400 pixels
+# Requires ImageMagick to be installed
```
Comment (mxochicale) [inline data/bash-scripts/resizing_images_in_folder.bash:4]: Nice one :)

File: src/ready/apis/train_mobious.py [hunk 4]
```diff
@@ -220,25 +225,40 @@ def main(args):
...
     target_transform_arg = transform_map.get(config.transforms.target_transform_operation, None)
-
+    
+
+    # Split datapath into trainpath, validation+path, test_path
+    # Create full mobious dataset first, then split into respective sets.
+    # Make sure to change dataset fed into trainloader
+    # use .random_split()
+    # Put validation into training loop
+    # Finally, evaluate with the test_set
```
Comment (mxochicale) [inline src/ready/apis/train_mobious.py:235]: Either remove comments or move them to relevant docstring

File: src/ready/apis/train_mobious.py [hunk 5]
```diff
@@ -220,25 +225,40 @@ def main(args):
...
+    
     ## Length 5; github_data_path
     ## Length 1143;  data_path
-    trainset = MobiousDataset(
+    full_dataset = MobiousDataset(
         data_path, transform=transform_arg ,target_transform=target_transform_arg
         )
+    
+    SEED = 42
```
Comment (mxochicale) [inline src/ready/apis/train_mobious.py:243]: Can add details in the doctring why we are using SEED 42? Should we use a specific SEED value? does it work with any value?

File: src/ready/apis/train_mobious.py [hunk 6]
```diff
@@ -326,17 +372,74 @@ def main(args):
...
+        
+        print(f"Training Metrics:")
+        for key in performance_metrics_labels: 
+            training_performance[key] /= num_samples
+            print(f"Average {key} @ epoch: {training_performance[key]:.4f}")
+        
+        print(f"\nValidation Metrics:")
+        for key in performance_metrics_labels:     
+            validation_performance[key] /= num_samples
+            print(f"Average {key} @ epoch: {validation_performance[key]: .4f}")
```
Comment (mxochicale) [inline src/ready/apis/train_mobious.py:439]: can we mention how to plot training and validation metrics as in plot losses at https://github.com/oocular/ready/tree/93-splitting-data-into-sets/data/mobious/models. Maybe also you can add descriptions within the [README](https://github.com/oocular/ready/blob/93-splitting-data-into-sets/data/mobious/models/README.md)

**Review by mxochicale — CHANGES_REQUESTED** [review]
Hi @stoppedtime24 

Looking great. Please have look to few commends I made and let me know what you think. 

Thanks, Miguel

File: data/bash-scripts/resizing_images_in_folder.bash (hunk 1 above)
Comment (stoppedtime24) [reply data/bash-scripts/resizing_images_in_folder.bash]: Addressed in 39a8076.

File: src/ready/apis/train_mobious.py [hunk 7]
```diff
@@ -220,25 +225,40 @@ def main(args):
...
     ## Length 1143;  data_path
-    trainset = MobiousDataset(
+    full_dataset = MobiousDataset(
         data_path, transform=transform_arg ,target_transform=target_transform_arg
         )
+    
+    SEED = 42
+    train_set, validation_set, test_set = torch.utils.data.random_split(full_dataset, 
+                                                                        [0.70, 0.15, 0.15],
+                                                                        torch.Generator().manual_seed(SEED))
```
[the remark this answers is not in the document]
Comment (stoppedtime24) [reply src/ready/apis/train_mobious.py]: Addressed in 39a8076. Created some variables that reference the proportion arguments in the config file.

File: configs/models/unet/config_train_unet_with_mobious.yaml [hunk 8]
```diff
@@ -18,6 +18,11 @@ transforms: # Transformations are defined in train_mobious.py using transforms.C
     transform_operation: None # Options are 'transforms_img', 'transforms_rotations', None
     target_transform_operation: None # Options are 'transforms_img', 'transforms_rotations', None
 
+datasets_proportions: # Ensure sum of all proportions = 1.
```
Comment (mxochicale) [inline configs/models/unet/config_train_unet_with_mobious.yaml:21]: I believe dataset_splitting_ratios is commonly used terminology, even though 'proportions' means the same thing. So just rename datasets_proportions with `dataset_splitting_ratios`

File: src/ready/apis/train_mobious.py [hunk 9]
```diff
@@ -226,23 +228,20 @@ def main(args):
...
     
     SEED = 42
+
+    TRAIN_SET_PROPORTION = config.datasets_proportions.train_set
+    VALIDATION_SET_PROPORTION = config.datasets_proportions.validation_set
+    TEST_SET_PROPORTION = config.datasets_proportions.test_set
+
     train_set, validation_set, test_set = torch.utils.data.random_split(full_dataset, 
-                                                                        [0.70, 0.15, 0.15],
+                                                                        [TRAIN_SET_PROPORTION, VALIDATION_SET_PROPORTION, TEST_SET_PROPORTION],
```
Comment (mxochicale) [inline src/ready/apis/train_mobious.py:244]: DITTO. let's use `..._RATIO` for these variables as it is commonly used terminology.

**Review by mxochicale — APPROVED** [review]
Thanks @stoppedtime24 

It looks good to me. I have just added a few comments, please take a look and address them. Feel free to squash and merge afterwards.

Thanks, Miguel
